### PR TITLE
refactor: Ziofy ProjectsResponderADM DEV-1728

### PR DESCRIPTION
**Pull request by seakayone**
### Task Description/Number

Issue Number: DEV-1728

Migrates the `ProjectsResponderADM` into ZIO land.

* Introduces `ProjectsResponderADM` trait which exposes the methods from the receive/handle to the outside world for possible direct use
* Adds a live implementation and layer for above trait
* Adds a ZIO implementation for the `IriLocker`
* Adds a `ZioHelper` object which implements functions for `ZIO`s akin to Scala's `Future#sequence` functions
* Adds overloaded `Triplestore#sparqlHttpExtendedConstruct` methods which accepts a `String`
* Fixes handling of errors in `ActorUtil#zio2Message`, all causes are sent back to the actor as failure

PR is based on https://github.com/dasch-swiss/dsp-api/pull/2453 that is why I have set this as the merge target for now so that you do not have to review the other PR twice.

### Basic Requirements

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### PR Type

What kind of change does this PR introduce?

- [ ] fix: represents bug fixes
- [x] refactor: represents production code refactoring
- [ ] feat: represents a new feature
- [ ] docs: documentation changes (no production code change)
- [ ] chore: maintenance tasks (no production code change)
- [ ] test: all about tests: adding, refactoring tests (no production code change)
- [ ] other... Please describe:

### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No
- [ ] Maybe (not 100% sure => check with FE)

### Does this PR change client-test-data?

- [ ] Yes (don't forget to update the JS-LIB team about the change)
- [x] No

### CODE REVIEW

**Review by subotic — APPROVED** [review]

**Review by BalduinLandolt — APPROVED** [review]
LGTM, just one minor suggestion, feel free to ignore it

File: webapi/src/main/scala/org/knora/webapi/responders/IriService.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  def makeUnusedIri(iriFun: => IRI): Task[IRI] = {
+    def makeUnusedIriRec(attempts: Int): Task[IRI] = {
+      val newIri = iriFun
+      for {
+        iriExists <- checkIriExists(newIri)
+        result <- if (!iriExists) {
+                    ZIO.succeed(newIri)
+                  } else if (attempts == 0) {
+                    ZIO.fail(
+                      UpdateNotPerformedException(s"Could not make an unused new IRI after $MAX_IRI_ATTEMPTS attempts")
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/responders/IriService.scala:73]:
```suggestion
                      UpdateNotPerformedException(s"Could not make an unused new IRI after $MAX_IRI_ATTEMPTS attempts.")
```

File: webapi/src/main/scala/org/knora/webapi/responders/IriService.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                 )
+               )
+
+        } yield entityIriAsString
+
+      case None => makeUnusedIri(iriFormatter)
+    }
+
+  def makeUnusedIri(iriFun: => IRI): Task[IRI] = {
+    def makeUnusedIriRec(attempts: Int): Task[IRI] = {
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/responders/IriService.scala:65]: Is it `@tailrec`?

File: webapi/src/main/scala/org/knora/webapi/responders/Responder.scala [hunk 3]
```diff
@@ -42,9 +45,15 @@ abstract class Responder(actorDeps: ActorDeps) extends LazyLogging {
...
+    FastFuture.failed(unexpectedMessageException(message: Any, who: IRI))
+}
+
+object Responder {
+
+  def handleUnexpectedMessage(message: Any, who: String): Task[Nothing] =
+    ZIO.fail(unexpectedMessageException(message, who))
+  private def unexpectedMessageException(message: Any, who: IRI) =
+    UnexpectedMessageException(
+      s"$who received an unexpected message $message of type ${message.getClass.getCanonicalName}"
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/responders/Responder.scala:57]:
```suggestion
      s"$who received an unexpected message $message of type ${message.getClass.getCanonicalName}."
```

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala [hunk 4]
```diff
@@ -43,16 +43,164 @@ import org.knora.webapi.messages.util.rdf._
...
+   * @return all the projects as a [[ProjectADM]].
+   *
+   *         NotFoundException if no projects are found.
+   */
+  def projectsGetRequestADM(): Task[ProjectsGetResponseADM]
+
+  /**
+   * Gets the project with the given project IRI, shortname, shortcode or UUID and returns the information as a [[ProjectADM]].
+   *
+   * @param identifier the IRI, shortname, shortcode or UUID of the project.
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala:70]:
```suggestion
   * Gets the project with the given project IRI, shortname or shortcode and returns the information as a [[ProjectADM]].
   *
   * @param identifier the IRI, shortname or shortcode of the project.
```

There is no UUID identifier anymore. It was living only during the "make project IRI not a shortcode but UUID based" crusade. I just forgot to remove it from docstrings here and there or someone brought it back with bad conflicts merge. You can ignore this suggestion here, I can fix it in separate PR.

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala [hunk 5]
```diff
@@ -43,16 +43,164 @@ import org.knora.webapi.messages.util.rdf._
...
+   *         [[BadRequestException]]     In the case when the shortcode is invalid.
+   */
+  def projectCreateRequestADM(
+    createProjectRequest: ProjectCreatePayloadADM,
+    requestingUser: UserADM,
+    apiRequestID: UUID
+  ): Task[ProjectOperationResponseADM]
+
+  /**
+   * Update project's basic information.
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala:172]:
```suggestion
   * Updates project's basic information.
```

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala [hunk 6]
```diff
@@ -161,13 +307,14 @@ final case class ProjectsResponderADM(actorDeps: ActorDeps, cacheServiceSettings
...
+   *         NotFoundException if no projects are found.
    */
-  private def projectsGetRequestADM(): Future[ProjectsGetResponseADM] =
+  override def projectsGetRequestADM(): Task[ProjectsGetResponseADM] =
     for {
       projects <- projectsGetADM()
-      result = if (projects.nonEmpty) { ProjectsGetResponseADM(projects = projects) }
-               else { throw NotFoundException(s"No projects found") }
+      result <- if (projects.nonEmpty) { ZIO.succeed(ProjectsGetResponseADM(projects)) }
+                else { ZIO.fail(NotFoundException(s"No projects found")) }
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala:317]:
```suggestion
                else { ZIO.fail(NotFoundException("No projects found.")) }
```

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala [hunk 7]
```diff
@@ -177,47 +324,38 @@ final case class ProjectsResponderADM(actorDeps: ActorDeps, cacheServiceSettings
...
+          getProjectFromTriplestore(identifier = identifier)
+        } else {
+          // getting from cache or triplestore
+          getProjectFromCacheOrTriplestore(identifier = identifier)
+        }
+    } yield maybeProjectADM
+  }
 
   /**
    * Gets the project with the given project IRI, shortname, shortcode or UUID and returns the information
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala:350]: Here is the same story with `UUID` for all the methods but easier would be to use replace tool. Feel free to do it or we can address it somewhere else.

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala [hunk 8]
```diff
@@ -989,43 +1049,42 @@ final case class ProjectsResponderADM(actorDeps: ActorDeps, cacheServiceSettings
...
-        shortnameExists <- projectByShortnameExists(createProjectRequest.shortname.value)
-        _ = if (shortnameExists) {
-              throw DuplicateValueException(
-                s"Project with the shortname: '${createProjectRequest.shortname.value}' already exists"
-              )
-            }
+        _ <- ZIO
+               .fail(
+                 DuplicateValueException(
+                   s"Project with the shortname: '${createProjectRequest.shortname.value}' already exists"
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala:1063]:
```suggestion
                   s"Project with the shortname: '${createProjectRequest.shortname.value}' already exists."
```

File: webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala [hunk 9]
```diff
@@ -989,43 +1049,42 @@ final case class ProjectsResponderADM(actorDeps: ActorDeps, cacheServiceSettings
...
-
-        _ = if (shortcodeExists) {
-              throw DuplicateValueException(
-                s"Project with the shortcode: '${createProjectRequest.shortcode.value}' already exists"
-              )
-            }
+        _ <- ZIO
+               .fail(
+                 DuplicateValueException(
+                   s"Project with the shortcode: '${createProjectRequest.shortcode.value}' already exists"
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/responders/admin/ProjectsResponderADM.scala:1072]:
```suggestion
                   s"Project with the shortcode: '${createProjectRequest.shortcode.value}' already exists."
```

File: webapi/src/main/scala/org/knora/webapi/routing/UnsafeZioRun.scala [hunk 10]
```diff
@@ -21,6 +21,16 @@ object UnsafeZioRun {
   def run[R, E, A](effect: ZIO[R, E, A])(implicit r: Runtime[R]): Exit[E, A] =
     Unsafe.unsafe(implicit u => r.unsafe.run(effect))
 
+  /**
+   * Executes the effect synchronously and returns its result as a
+   * [[A]] value or throws a [[FiberFailure]].
+   *
+   * This method is effectful and should only be used at the edges of your
```
Comment (mpro7) [inline webapi/src/main/scala/org/knora/webapi/routing/UnsafeZioRun.scala:28]: I guess `effectful` in terms of ZIO, but could you be more specific here maybe?

**Review by mpro7 — APPROVED** [review]
Nice one. Just please answer my questions and check cosmetic suggestions:

File: webapi/src/main/scala/org/knora/webapi/responders/IriService.scala (hunk 2 above)
Comment (seakayone) [reply webapi/src/main/scala/org/knora/webapi/responders/IriService.scala:65]: No.

**Pull request merged**